pkg/parser/expression.go

```diff
@@ -0,0 +1,402 @@
+	return exp
+}
+
+func (p *Parser) parserIndexExpr(scope *scope, left Node) Node {
+	tok := p.cur // TODO: ensure not prefixed by WS
+	p.advance()  // advance past [
+	leftType := left.Type().Name
+	if leftType != ARRAY && leftType != MAP && leftType != STRING {
+		p.appendErrorForToken("only array, string and map type can be indexed, found "+left.Type().Format(), tok)
+		return nil
+	}
+	index := p.parseTopLevelExpr(scope)
+	if !p.assertToken(lexer.RBRACKET) {
+		return nil
+	}
+	p.advance() // advance past ]
+	if leftType == ARRAY || leftType == STRING {
+		if index.Type() != NUM_TYPE {
+			p.appendError(leftType.String() + " index expects num, found " + index.Type().Format())
+			return nil
+		}
+		return &IndexExpression{Token: tok, Left: left, Index: index, T: left.Type().Sub}
+	}
+	if index.Type() != STRING_TYPE {
+		p.appendError("map index expects string, found " + index.Type().Format())
+		return nil
+	}
+	return &IndexExpression{Token: tok, Left: left, Index: index}
+}
+
+func (p *Parser) parserDotExpr(left Node) Node {
+	tok := p.cur // TODO: ensure not prefixed by WS
+	p.advance()  // advance past .
+	leftType := left.Type().Name
+	if leftType != MAP {
+		p.appendErrorForToken("field access with '.' expects map type, found "+left.Type().Format(), tok)
+		return nil
+	}
+	if p.cur.TokenType() != lexer.IDENT {
+		p.appendErrorForToken("expected map key, found "+p.cur.TokenType().Format(), tok)
+		return nil
+	}
+	expr := &DotExpression{Token: tok, Left: left, T: left.Type().Sub, Key: p.cur.Literal}
+	p.advance() // advance past key IDENT
+	return expr
+}
+
+func isBinaryOp(tt lexer.TokenType) bool {
+	return isComparisonOp(tt) || tt == lexer.PLUS || tt == lexer.MINUS || tt == lexer.SLASH || tt == lexer.ASTERISK || tt == lexer.OR || tt == lexer.AND
+}
+
+func isComparisonOp(tt lexer.TokenType) bool {
+	return tt == lexer.EQ || tt == lexer.NOT_EQ || tt == lexer.LT || tt == lexer.GT || tt == lexer.LTEQ || tt == lexer.GTEQ
+}
+
+func (p *Parser) validateUnaryType(unaryExp *UnaryExpression) {
+	tok := unaryExp.Token
+	op := unaryExp.Op
+	if op != OP_MINUS && op != OP_BANG {
+		p.appendErrorForToken("invalid unary operator", tok)
+		return
+	}
+	rightType := unaryExp.Right.Type()
+	if op == OP_MINUS {
+		if rightType != NUM_TYPE {
+			p.appendErrorForToken("'-' unary expects num type, found "+rightType.String(), tok)
+		}
+		return
+	}
+	if rightType != BOOL_TYPE {
+		p.appendErrorForToken("'!' unary expects bool type, found "+rightType.String(), tok)
+	}
+}
+
+func (p *Parser) validateBinaryType(binaryExp *BinaryExpression) {
+	tok := binaryExp.Token
+	op := binaryExp.Op
+	if op == OP_ILLEGAL || op == OP_BANG {
+		p.appendErrorForToken("invalid binary operator", tok)
+		return
+	}
+
+	leftType := binaryExp.Left.Type()
+	rightType := binaryExp.Right.Type()
+	if !leftType.Equals(rightType) {
+		p.appendErrorForToken("mismatched type for "+op.String()+": "+leftType.Format()+", "+rightType.Format(), tok)
+		return
+	}
+	if op == OP_PLUS {
```

Comment:
i think i'd use a switch again - then you don't need the `return` statements. it would look a lot like the unary validation

Reply:
yeah agreed, looks a lot better.